<a href="https://colab.research.google.com/github/ManuSinghYadav/Mushroom_Classification_All_Models/blob/main/Mushroom_Classification_(All_models).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Loading the data
Dataset can be downloaded from [here](https://www.kaggle.com/datasets/uciml/mushroom-classification).

In [36]:
from google.colab import files
upload = files.upload()

Saving mushrooms data.csv to mushrooms data.csv


In [37]:
import pandas as pd
a=pd.read_csv('mushrooms data.csv')

##Preprocessing & Basic Analysis

In [40]:
a.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [41]:
a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [42]:
a.isnull().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [43]:
a.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [45]:
a.shape

(8124, 23)

In [86]:
x = a.drop(['class'], axis=1)
y = a['class']

####Encoding the Categorical columns (**OneHotEncoder**)

In [90]:
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder()
x_transformed = ohe.fit_transform(x)

In [91]:
x_transformed

<8124x117 sparse matrix of type '<class 'numpy.float64'>'
	with 178728 stored elements in Compressed Sparse Row format>

As this is in "sparse matrix", we will transform it into an array and then to a dataframe.

In [92]:
x_transformed_dense_array = x_transformed.toarray()

In [93]:
x_transformed = pd.DataFrame(x_transformed_dense_array)

In [94]:
x_transformed

0    1    2    3    4    5    6    7    8    9    ...  107  108  109  \
0     0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  ...  1.0  0.0  0.0   
1     0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0   
2     1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0   
3     0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  ...  1.0  0.0  0.0   
4     0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
8119  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0   
8120  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  ...  0.0  1.0  0.0   
8121  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0   
8122  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  1.0  0.0   
8123  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0   

      110  111  112  113  114  115  116  
0     0.0  0.0  0.0  0.0  0.0  1.0  0.0  
1     0.0  1.0  0.0  0.0  0.0  0.0  0.0  
2     0.0  0.0  0.0  1.0  0.0  0.0  0.0  
3     0.0  0.0  0.0  0.0  0.0  1.0  0.0  
4     0.0  1.0  0.0  0.0  0.0  0.0  0.0  
...   ...  ...  ...  ...  ...  ...  ...  
8119  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8120  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8121  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8122  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8123  0.0  0.0  1.0  0.0  0.0  0.0  0.0  

[8124 rows x 117 columns]

###Balancing the dataset.
As the dataset is imbalenced. Lets balence it by using **SMOTE** technique.

In [97]:
y.value_counts()

0    4208
1    3916
Name: class, dtype: int64

In [100]:
from imblearn.over_sampling import SMOTE

S = SMOTE()
os_x, os_y = S.fit_resample(x_transformed, y)

#Here os represents "over-sampled"

In [105]:
os_y.value_counts()

1    4208
0    4208
Name: class, dtype: int64

In [104]:
os_x.shape

(8416, 117)

###**Feature Selection**
As now we have got 117 features (columns). Let's reduce them by using **RFE** technique.

In [106]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression

rfe = RFE(estimator = LogisticRegression())
rfe = rfe.fit(os_x, os_y)

In [110]:
os_x.columns[rfe.support_]

Int64Index([  1,   4,   6,   7,  10,  11,  15,  18,  22,  23,  24,  25,  27,
             28,  29,  30,  33,  34,  35,  36,  37,  38,  45,  49,  50,  51,
             52,  53,  55,  56,  57,  58,  60,  62,  63,  66,  69,  72,  75,
             77,  78,  81,  86,  90,  91,  92,  96,  97,  98, 100, 101, 105,
            106, 109, 110, 111, 113, 116],
           dtype='int64')

In [112]:
len(os_x.columns[rfe.support_])

58

In [117]:
x = os_x[os_x.columns[rfe.support_]]

In [118]:
x

1    4         6    7    10   11   15   18   22   23   ...  98   100  \
0     0.0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
1     0.0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  1.0  0.0   
2     0.0  0.0  0.000000  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  1.0  0.0   
3     0.0  0.0  0.000000  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0   
4     0.0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0   
...   ...  ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
8411  0.0  0.0  1.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8412  0.0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8413  0.0  0.0  0.220989  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8414  0.0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8415  0.0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   

      101  105  106  109  110       111  113  116  
0     0.0  0.0  0.0  0.0  0.0  0.000000  0.0  0.0  
1     0.0  0.0  1.0  0.0  0.0  1.000000  0.0  0.0  
2     0.0  0.0  1.0  0.0  0.0  0.000000  1.0  0.0  
3     0.0  0.0  0.0  0.0  0.0  0.000000  0.0  0.0  
4     0.0  0.0  0.0  0.0  0.0  1.000000  0.0  0.0  
...   ...  ...  ...  ...  ...       ...  ...  ...  
8411  0.0  0.0  0.0  0.0  0.0  0.671444  0.0  0.0  
8412  0.0  0.0  0.0  0.0  0.0  0.000000  0.0  0.0  
8413  0.0  0.0  0.0  0.0  0.0  0.000000  0.0  0.0  
8414  0.0  0.0  0.0  0.0  0.0  0.000000  0.0  0.0  
8415  0.0  0.0  0.0  0.0  0.0  0.000000  0.0  0.0  

[8416 rows x 58 columns]

###**Train-Test Split**

In [120]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, os_y, test_size=0.25)

##Training (ML Models)

###1. Logistic Regression

In [121]:
from sklearn.linear_model import LogisticRegression

LR=LogisticRegression().fit(x_train,y_train)

###2. K-Nearest Neighbor (K-NN)

In [122]:
from sklearn.neighbors import KNeighborsClassifier

KNN = KNeighborsClassifier(n_neighbors=5).fit(x_train, y_train)

###3. Support Vector Machine (SVM)

In [123]:
from sklearn.svm import SVC

SVM = SVC(kernel='linear').fit(x_train, y_train)

###4. Naive Bayes

In [124]:
from sklearn.naive_bayes import GaussianNB

GNB = GaussianNB().fit(x_train, y_train)

###5. Decision Tree

In [127]:
from sklearn.tree import DecisionTreeClassifier

DT = DecisionTreeClassifier().fit(x_train, y_train)

###6. Random Forest (Bagging)

In [128]:
from sklearn.ensemble import RandomForestClassifier

RF = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0).fit(x_train, y_train)

###7. XGBoost (Boosting)

In [139]:
from xgboost import XGBClassifier

XGB = XGBClassifier().fit(x_train, y_train)

##Testing / Model Scores

In [140]:
print("Model Scroes,\n1. Logistic Regression: {}\n2. KNN: {}\n3. SVM: {}\n4. Naive Bayes: {}\n5. Decision Tree: {}\n6. Random Forest: {}\n7. XGBoost: {}".format(LR.score(x_test,y_test), KNN.score(x_test,y_test), SVM.score(x_test,y_test), GNB.score(x_test,y_test), DT.score(x_test,y_test), RF.score(x_test,y_test), XGB.score(x_test,y_test)))

Model Scroes,
1. Logistic Regression: 1.0
2. KNN: 1.0
3. SVM: 1.0
4. Naive Bayes: 0.9928707224334601
5. Decision Tree: 1.0
6. Random Forest: 1.0
7. XGBoost: 1.0
